RecoLocalCalo/EcalRecProducers/plugins/EcalRecHitConvertGPU2CPUFormat.cc

```diff
@@ -0,0 +1,124 @@
+// framework
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+// algorithm specific
+#include "DataFormats/EcalDigi/interface/EcalDigiCollections.h"
+#include "CUDADataFormats/EcalRecHitSoA/interface/EcalRecHit_soa.h"
+#include "DataFormats/EcalRecHit/interface/EcalRecHit.h"
+#include "DataFormats/EcalRecHit/interface/EcalRecHitCollections.h"
+#include "RecoLocalCalo/EcalRecAlgos/interface/Common.h"
+
+#include <iostream>
+
+class EcalRecHitConvertGPU2CPUFormat : public edm::stream::EDProducer<> {
+public:
+  explicit EcalRecHitConvertGPU2CPUFormat(edm::ParameterSet const& ps);
+  ~EcalRecHitConvertGPU2CPUFormat() override;
+  static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+private:
+  using GPURecHitType = ecal::RecHit<ecal::Tag::soa>;
+  void produce(edm::Event&, edm::EventSetup const&) override;
+
+private:
+  const edm::EDGetTokenT<ecal::SoARecHitCollection> recHitsGPUEB_;
+  const edm::EDGetTokenT<ecal::SoARecHitCollection> recHitsGPUEE_;
+
+  const std::string recHitsLabelCPUEB_, recHitsLabelCPUEE_;
+};
+
+void EcalRecHitConvertGPU2CPUFormat::fillDescriptions(edm::ConfigurationDescriptions& confDesc) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("recHitsLabelGPUEB", edm::InputTag("ecalRecHitProducerGPU", "EcalRecHitsGPUEB"));
+  desc.add<edm::InputTag>("recHitsLabelGPUEE", edm::InputTag("ecalRecHitProducerGPU", "EcalRecHitsGPUEE"));
+
+  desc.add<std::string>("recHitsLabelCPUEB", "EcalRecHitsEB");
+  desc.add<std::string>("recHitsLabelCPUEE", "EcalRecHitsEE");
+
+  std::string label = "ecalRecHitConvertGPU2CPUFormat";
+  confDesc.add(label, desc);
+}
+
+EcalRecHitConvertGPU2CPUFormat::EcalRecHitConvertGPU2CPUFormat(const edm::ParameterSet& ps)
+    : recHitsGPUEB_{consumes<ecal::SoARecHitCollection>(ps.getParameter<edm::InputTag>("recHitsLabelGPUEB"))},
+      recHitsGPUEE_{consumes<ecal::SoARecHitCollection>(ps.getParameter<edm::InputTag>("recHitsLabelGPUEE"))},
+      recHitsLabelCPUEB_{ps.getParameter<std::string>("recHitsLabelCPUEB")},
+      recHitsLabelCPUEE_{ps.getParameter<std::string>("recHitsLabelCPUEE")} {
+  produces<EBRecHitCollection>(recHitsLabelCPUEB_);
+  produces<EERecHitCollection>(recHitsLabelCPUEE_);
+}
+
+EcalRecHitConvertGPU2CPUFormat::~EcalRecHitConvertGPU2CPUFormat() {}
+
+void EcalRecHitConvertGPU2CPUFormat::produce(edm::Event& event, edm::EventSetup const& setup) {
+  edm::Handle<ecal::SoARecHitCollection> hRecHitsGPUEB, hRecHitsGPUEE;
+  event.getByToken(recHitsGPUEB_, hRecHitsGPUEB);
+  event.getByToken(recHitsGPUEE_, hRecHitsGPUEE);
```

Comment:
```suggestion
  auto const& hRecHitsGPUEB = event.get(recHitsGPUEB_);
  auto const& hRecHitsGPUEE = event.get(recHitsGPUEE_);
```
would be more efficient, each re-reference of an `edm::Handle<T>` checks if the product `T` is valid.

Reply:
fixed in next commit